### PR TITLE
Fixed #3841 - Saving profile deletes SMTP password

### LINKED ISSUE #3841: Saving profile deletes SMTP password
#### Issue

On a system configured to not allow users to use the system's outbound SMTP account:
Saving my user profile, deletes my own SMTP password.

The problem is that when editing the profile, the password field is by default visible, editable and blank.
Even if the field is not touched in any way, when submiting the form, the CRM thinks the password was removed.

This password should behave like the one in the system outbound email settings: display a change password "link".

##### Profile:
screenshot from 2017-07-05 18 06 00

##### System settings:
screenshot from 2017-07-05 18 14 44

#### Expected Behavior

Saving my profile, should not reset my SMTP password.
The password field should have a javascript link to enable changing the password, like in the system outbound email settings:

#### Actual Behavior

Saving my profile, resets my SMTP password even if I didn't touch the password field.

#### Steps to Reproduce

1. In the outbound email settings, use an SMTP account and disable its use by users
2. Go to your profile, add a SMTP password and save
3. Check that a password exists for your account in the outbound_email mysql table
4. Save your profile again
5. Check the mysql table again. An empty string is in the `mail_smtppass` field.

#### Your Environment

* SuiteCRM Version used: 7.9.2
* Browser name and version: Chromium 58
* Environment name and version: MySQL 5.7, PHP 7.0.18, Nginx 1.10
* Operating System and version: Ubuntu 16.04

### PR DESCRIPTION
**Pull request by daniel-samson**
## Description

Users are able to change the system outbound email account by editing it in their user profile.

## Motivation and Context

* This change removes there ability to change the password in the user profile.

## How To Test This

* Repair and Rebuild
* Access user profile
* Attempt to change the system email accounts password

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist

- [x] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [ ] My change requires a change to the documentation.
- [x] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php/Contributing_to_SuiteCRM) guidelines.

### CODE REVIEW

File: include/OutboundEmail/OutboundEmail.php [hunk 1]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
         		      $autoCreateUserSystemOverride = TRUE;
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: PSR-2:
"The PHP constants true, false, and null MUST be in lower case."

File: include/OutboundEmail/OutboundEmail.php [hunk 2]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
-								   'is_editable' => $isEditable, 'type' => $system->type, 'errors' => $systemErrors);
+                // Substitute in the users system override if its available.
+                if($userSystemOverride != null) {
+                    $system = $userSystemOverride;
+                }
+        		else if ($autoCreateUserSystemOverride) {
+                    $system = $this->createUserSystemOverrideAccount($user->id, "", "");
+                }
+                // User overrides can be edited.
+                $isEditable = ($system->type == 'system' || $system->type == 'system-override') ? FALSE : TRUE;
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: PSR-2:
"The PHP constants true, false, and null MUST be in lower case."

File: include/OutboundEmail/OutboundEmail.php [hunk 3]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
-			    $isEditable = ($system->type == 'system') ? FALSE : TRUE; //User overrides can be edited.
-
-                if( !empty($system->mail_smtpserver) )
-				    $ret[] = array('id' =>$system->id, 'name' => "$system->name", 'mail_smtpserver' => $system->mail_smtpdisplay,
-								   'is_editable' => $isEditable, 'type' => $system->type, 'errors' => $systemErrors);
+                // Substitute in the users system override if its available.
+                if($userSystemOverride != null) {
+                    $system = $userSystemOverride;
+                }
+        		else if ($autoCreateUserSystemOverride) {
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: wrong indent

File: include/OutboundEmail/OutboundEmail.php [hunk 4]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
+                        'id' => $system->id,
+                        'name' => "$system->name",
+                        'mail_smtpserver' => $system->mail_smtpdisplay,
+                        'is_editable' => $isEditable,
+                        'type' => $system->type,
+                        'errors' => $systemErrors
+                    );
+                }
 			}
 			else //Sendmail
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: PSR-2:
"else and elseif are on the same line as the closing brace from the earlier body."

File: include/OutboundEmail/OutboundEmail.php [hunk 5]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
+                    $ret[] = array(
+                        'id' => $system->id,
+                        'name' => "$system->name",
+                        'mail_smtpserver' => $system->mail_smtpdisplay,
+                        'is_editable' => $isEditable,
+                        'type' => $system->type,
+                        'errors' => $systemErrors
+                    );
+                }
 			}
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: wrong indent

File: include/OutboundEmail/OutboundEmail.php [hunk 6]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
 			else //Sendmail
 			{
-				$ret[] = array('id' =>$system->id, 'name' => "{$system->name} - sendmail", 'mail_smtpserver' => 'sendmail',
-								'is_editable' => false, 'type' => $system->type, 'errors' => '');
+				$ret[] = array(
+				    'id' =>$system->id,
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: wrong indent

File: include/OutboundEmail/OutboundEmail.php [hunk 7]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
+                    'name' => "{$system->name} - sendmail",
+                    'mail_smtpserver' => 'sendmail',
+					'is_editable' => false,
+                    'type' => $system->type,
+                    'errors' => ''
+                );
 			}
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: wrong indent

File: include/OutboundEmail/OutboundEmail.php [hunk 8]
```diff
@@ -199,40 +199,56 @@ function getUserMailers($user) {
...
 		}
 
 		while($a = $this->db->fetchByAssoc($r))
 		{
 			$oe = array();
-			if($a['mail_sendtype'] != 'SMTP')
+			if($a['mail_sendtype'] != 'SMTP') {
 				continue;
-
+            }
```
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: wring indent

File: modules/Users/UserViewHelper.php [hunk 9]
```diff
@@ -757,8 +757,6 @@ protected function setupEmailSettings() {
             $this->ss->assign('MAIL_SMTPPORT',$mail_smtpport);
             $this->ss->assign('MAIL_SMTPSSL',$mail_smtpssl);
         }
-        $this->ss->assign('HIDE_IF_CAN_USE_DEFAULT_OUTBOUND',$hide_if_can_use_default );
```
Comment (gymad) [inline modules/Users/UserViewHelper.php]: Users can not use default outbound anymore but this calculation still there in file 
modules/Users/views/view.wizard.php at class ViewWizard::display()

File: modules/Users/UserViewHelper.php [hunk 10]
```diff
@@ -757,8 +757,6 @@ protected function setupEmailSettings() {
             $this->ss->assign('MAIL_SMTPPORT',$mail_smtpport);
             $this->ss->assign('MAIL_SMTPSSL',$mail_smtpssl);
         }
```
Comment (gymad) [inline modules/Users/UserViewHelper.php]: unused $hide_if_can_use_default variable:
Users can not use default outbound anymore but still calculate this value before this part.

File: include/OutboundEmail/OutboundEmail.php (hunk 2 above)
Comment (pgorod) [inline include/OutboundEmail/OutboundEmail.php]: `$isEditable = ($system->type == 'system' || $system->type == 'system-override') ? FALSE : TRUE;`
is the same as simply
`$isEditable = !($system->type == 'system' || $system->type == 'system-override')`
which is the same as 
`$isEditable = ($system->type != 'system' && $system->type != 'system-override')`
... I would use this last one, but use whichever you prefer :-)

File: modules/Users/tpls/EditViewFooter.tpl [hunk 11]
```diff
@@ -78,39 +78,6 @@
                         </select>
                     </td>
                 </tr>
-                            {if !$HIDE_IF_CAN_USE_DEFAULT_OUTBOUND}
```
Comment (gymad) [inline modules/Users/tpls/EditViewFooter.tpl]: $HIDE_IF_CAN_USE_DEFAULT_OUTBOUND template variable still there in:
/var/www/html/SuiteCRM1/modules/Emails/templates/emailOptions.tpl
/var/www/html/SuiteCRM1/modules/Users/UserEmailOptions.tpl
/var/www/html/SuiteCRM1/modules/Users/tpls/wizard.tpl
is that correct?

File: include/OutboundEmail/OutboundEmail.php (hunk 2 above)
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: $isEditable = !in_array('system', 'system-override')
;) :D

**Comment by daniel-samson** [comment]
Updated Pull Request to meet code review.

**Comment by gymad** [comment]
$HIDE_IF_CAN_USE_DEFAULT_OUTBOUND template var is still there in /var/www/html/SuiteCRM1/modules/Users/tpls/wizard.tpl

File: include/OutboundEmail/OutboundEmail.php (hunk 2 above)
Comment (gymad) [inline include/OutboundEmail/OutboundEmail.php]: $isEditable = preg_match('/^system(-override){0,1}/')

**Pull request merged**